interfaces/builtin/cups.go

```diff
@@ -0,0 +1,61 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+// On systems where the slot is provided by an app snap, the cups interface is
+// the companion interface to the cups-control interface. The design of these
+// interfaces is based on the idea that the slot implementation (eg cupsd) is
+// expected to query snapd on if the cups-control interface is connected or not
+// and the print service will mediate admin functionality (ie, the rules in
+// these interfaces allow connecting to the print service, but do not implement
+// enforcement rules; it is up to the print service to provide enforcement).
+const cupsSummary = `allows access to the CUPS socket for printing`
+
+// cups is currently only available via a providing app snap and this interface
+// assumes that the providing app snap also slots 'cups-control' (the current
+// design allows the snap provider to slots both cups-control and cups or just
+// cups-control (like with implicit classic or any slot provider without
+// mediation patches), but not just cups).
+const cupsBaseDeclarationSlots = `
```

Comment:
> not necessarily related to this PR, but is there any plan to make auto-connect by snaps that `plugs: [cups]` to auto-connect automatically to the "cups" providing snap? or are we not going to enable that, and just allow snaps to request (and grant) auto-connect for cups interface plugs to the cups snap cups slot (that's a lot of coffee smile )

Whether or not we auto-connect the cups interface will be entirely dependent on the slot provider which is why the base declaration is set the way it is, which uses the common pattern that by default, connect and auto-connect are disabled. If we do the vetting, etc, etc in forum requests for a slot provider, we'll grant connection. If it has the mediation patches, we'll grant auto-connection. The addition of the cups interface in this PR is in response to Till's work to add and upstream mediation patches to CUPS (so it is expected that the CUPS snap the slots 'cups' would allow auto-connect).

If those patches are backported to the CUPS debs in Ubuntu, we might consider auto-connection for implicit classic, like we do today with audio-playback (since most distros that support strict are forks/derivatives that pull from the Ubuntu archive and would therefore receive the mediation patches as well), but we can take a look at that if/when the mediation patches are SRU'd.